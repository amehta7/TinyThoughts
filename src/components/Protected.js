import React from 'react'
import { Route, Redirect } from 'react-router-dom'

const Protected = ({ user, children, ...rest }) => (
  <Route
    {...rest}
    render={({ location }) =>
      user ? (
        children
      ) : (
        <Redirect to={{ pathname: '/signin', state: { from: location } }} />
      )
    }
  />
)

export default Protected
