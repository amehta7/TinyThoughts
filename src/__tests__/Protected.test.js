import React from "react";
import { mount } from "enzyme";
import { MemoryRouter, Redirect as MockRedirect } from "react-router-dom";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import Protected from "../components/Protected";
const mockStore = configureStore([thunk]);

jest.mock("react-router-dom", () => {
  const RouterMocks = jest.requireActual("react-router-dom");
  return {
    ...RouterMocks,
    Redirect: jest.fn().mockImplementation(() => <span>REDIRECTED</span>)
  };
});

afterEach(() => {
  MockRedirect.mockClear();
});

describe("Protected Component", () => {
  it("Renders protected components when signed in", () => {
    const store = mockStore({
      users: {
        user: { id: 1, name: "Joe Mockery" }
      },
      posts: {
        categories: [
          { id: 1, title: "JavaScript" },
          { id: 2, title: "Python" }
        ]
      },
      errors: {
        error: false
      }
    });

    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter keyLength={0}>
          <Protected>
            <h1>DEMO CONTENT</h1>
          </Protected>
          >
        </MemoryRouter>
      </Provider>
    );

    return expect(wrapper.find(Protected)).toMatchSnapshot();
  });

  it("Redirects to the Signin page if signed out", () => {
    const store = mockStore({
      users: {},
      posts: {
        categories: [
          { id: 1, title: "JavaScript" },
          { id: 2, title: "Python" }
        ]
      },
      errors: {
        error: false
      }
    });

    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <Protected>
            <h1>DEMO CONTENT</h1>
          </Protected>
          >
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper.find(Protected).text()).toBe("REDIRECTED");
  });
});
