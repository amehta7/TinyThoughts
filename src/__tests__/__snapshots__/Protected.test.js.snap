// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Protected Component Renders protected components when signed in`] = `
<Protected>
  <Route
    render={[Function]}
  >
    <h1>
      DEMO CONTENT
    </h1>
  </Route>
</Protected>
`;

exports[`Protected Component Renders protected components when signed in 1`] = `
<Protected>
  <Route
    render={[Function]}
  >
    <mockConstructor
      to={
        Object {
          "pathname": "/signin",
          "state": Object {
            "from": Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
          },
        }
      }
    >
      <span>
        REDIRECTED
      </span>
    </mockConstructor>
  </Route>
</Protected>
`;
